### DISCUSSION TITLE
Way to allow only specific HTTP methods like GET/POST/PUT for a certain admin API endpoint

### DISCUSSION THREAD

**Discussion by sujaykulkarn** [keycloak/keycloak#26252] [Q&A]
Hi Community, 

I am trying to find a way to allow only specific HTTP methods for a certain admin API's endpoint, for better control of my application. Is there a way to configure the Keycloak.X to achieve this? 

I see currently all the HTTP methods are supported by Keycloak.X and a few sensitive information will also come with methods like HEAD and OPTION, so I want to restrict the usage of those. So if there are any possible options and suggestions to achieve this, please let me know, will try out all the options. 

Thanks

**Comment by danielFesenmeyer** [comment]
Afaik, there is no Keycloak builtin feature for that.

However, you could implement a Quarkus ServerRequestFilter as described here: https://quarkus.io/guides/resteasy-reactive#request-or-response-filters (Note that the "Jakarta REST way" described below does not work with Keycloak 23).

**Comment by sujaykulkarn** [comment]
@danielFesenmeyer Daniel, Thanks for the suggestion. Are you aware of any such code written just for my reference?
Thanks.

**Reply by danielFesenmeyer** [reply]
Don't know about other examples than the ones in the linked documentation. But those should be sufficient to get started, there is even an example for checking the used http method.

**Comment by darius-m** [comment]
You could also easily check the request method and end the connection early if a bad request is sent by using a reverse proxy. In nginx you could probably add something like
```nginx
if ($request_method !~ '^GET|POST|PUT|DELETE$') {
    return 444;
}
```

Do note, however, that you can run into issues if blocking valid requests. For example, the `OPTIONS` method is used in [Preflight requests in CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS#preflighted_requests_in_cors), so you may end up with some hard to debug issues. Unless you really know how requests are supposed to work, I think it would be safer to just block non-standard request methods (i.e., anything except `GET`, `POST`, `HEAD`, `OPTIONS`, `PUT`, `DELETE`), and rely on Quarkus' / Keycloak's routers for validation. `HEAD` may be a bit of a corner case among these, since I don't think I have seen it used for anything practical outside of testing connections with curl.
